### ISSUE TITLE
Fate unlimited codes. bug on multiplayer

### ISSUE THREAD

**Issue by isseihyoudoux** [hrydgard/ppsspp#13452]
### What happens? 
When the battle starts the game speed Up really fast on both players.

### What *should* happen?
the game should not be speeding Up fast when the battle starts on both players.

### What hardware, operating system, and PPSSPP version?  On desktop, GPU matters for graphical issues.

i used the last PPSSPP build v1.10.3-712-g7ed1ade56 everything works fine but. the problem is the multiplayer when the battle begins speeding up.

**Comment by MojoJojoDojo** [comment]
Bleach Heat The Soul series (tested only on 7 though) got the same problem as its created by the same developer (Eighting).

**Comment by ANR2ME** [comment]
It's strange for multiplayer to cause the game to renders at faster pace... i'm not familiar how PSP games normally do the rendering, may be these kind of games relies on networking to delays each frame? while the most common one on many games is using vsync to trigger the rendering.

For example, before the "Fight" initiated/begins the FPS/VPS is 30/30 on Bleach Heat the Soul 7, but after the "Fight" shows up it became 60/60 FPS/VPS which looks too fast.

**Comment by unknownbrackets** [comment]
Maybe callbacks are being registered and threads are being woken improperly or callbacks run on the wrong threads?  But definitely could also be that our network syscalls are too "fast" in emulated time.

I'd recommend benchmarking common calls used many times per frame on a PSP and comparing the results to PPSSPP with the same PSP code.  I've done this with a bunch of functions and it's both improved speed and timing.

-[Unknown]

**Comment by ANR2ME** [comment]
Btw i often saw `eatCycles` in HLE function, but never saw one on sceNet, is this eatCycles suppose to put delays on the HLE?
may be some sceNet functions need this eatCycles too (but not sure how many ticks)

i wonder if @adenovan ever benchmarks on sceNet functions, since he usually do auto-test on real PSP.

**Comment by Pik-a-boo** [comment]
Both the Bleach games and Fate both worn properly on JPCSP so maybe investigate and compare with it.

**Comment by unknownbrackets** [comment]
Yes, `hleDelayResult()` will block the calling PSP thread for that long, but doesn't affect the "progression of time".  The next thread will be scheduled immediately, as if the syscall took 0.00ms.

In contrast, `hleEatCycles()` (which can be used along with `hleDelayResult()`) consumes cycles, as if the syscall had to do a lot of work and took time.  Since the PSP is single-core for user threads, no other thread can run during this time and so that time is "consumed" or "eaten" by the syscall.  The next code (or thread, if rescheduled) will run after that progression of emulated time.

So for example, if you have PSP code like this (made up syscall):

```c
while (!sceNetReadFromSockNonBlocking(sock, buf, 1)) {
    continue;
}
```

Then, assuming it doesn't reschedule, you'll actually get really poor performance because only the CPU instructions will be consuming time.  If the actual theoretical `sceNetReadFromSockNonBlocking` takes 0.013ms per call, it'd be much better to consume that time, making the tight loop move time forward faster (and ultimately using less total mhz from the phone/computer.)

But an issue like that would typically just cause bad performance and low fps, not raise FPS.  It could still raise FPS if the game "knows" certain syscalls are expensive and assumes by calling a sequence of them, it'll already have lost 16ms so doesn't need to wait for that vsync.

-[Unknown]

**Comment by ANR2ME** [comment]
> 
> 
> Both the Bleach games and Fate work properly on JPCSP so maybe investigate and compare with it.
> https://youtube.com/watch?v=S6ERjsHDEnE&t=54s

Based on this video those games really are supposed to run at 30 FPS, but when i checked all 5 sceNet functions on JPCSP  that were used during the fight on Bleach 7, i couldn't find any kind of delaying mechanism on the non-blocking version

**Comment by ANR2ME** [comment]
LOL Fate Unlimited Codes had it worse, even the seconds countdown timer is running too fast.. i thought timers like these are being calculated using SystemTime/RTC or something instead of accumulating the delay on each loop

**Comment by MojoJojoDojo** [comment]
I also remember that one of the Burnout games also had that issue (dont remember which maybe both).

**Comment by ANR2ME** [comment]
@isseihyoudoux could you try the test build on PR https://github.com/hrydgard/ppsspp/pull/13519 to see if that PR really resolved this issue?

**Comment by somepunkid** [comment]
I'm testing this build now and the same issue seems to be present. If ran with Force real clock sync off, it runs at around 40fps, but the animation/game speed still seems fast. With it enabled, it is still having the 60fps/overspeed issue.

Let me know if a specific setting needs testing within this build. It's possible I don't have the workaround enabled? Or is it a just a compat flag?

**Comment by ANR2ME** [comment]
Yes, the Force real clock sync will also slowdown FPS, try disabling it. 
How about enabling slowdown network fps only? did you get 30 fps during battle? 
The characters intro might get lower fps (ie. 24 fps) because it's just a workaround, but during battle the fps should be stable at 30 fps right?

**Comment by somepunkid** [comment]
x64
Nvidia 2080 Max-Q  /Win10
Tested on same machine

Just did two fresh instances. I get 30fps at the start and for a round/for most of a around, but it will dip back down to 24fps between rounds at random and lock there. Will try another capable win10 machine to confirm.

**Comment by ANR2ME** [comment]
Ugh i guess this workaround aren't good, in my case i need to enable both force real clock sync and slowdown network fps in order to get 30 fps on both fate unimited codes and bleach heat the soul 7, without force real clock sync it goes back to 60 fps during battle :(

**Comment by somepunkid** [comment]
What's odd is, with this workaround, it will occasionally cap back at 30fps on the same stage even and stay that way for the round. Then, it will go back to 24fps. It is wildly inconsistent for me thus far.

**Comment by Pik-a-boo** [comment]
If someone just made 60 fps patches for both games this would have allivated the issue a bit.
Unless its impossible.

**Comment by somepunkid** [comment]
Just got done testing multiple stages and rounds with Force Clock and network slowdown on both machines. Can't replicate 30fps consistency on both sets of hardware, but sometimes multiple rounds stay synced to 30fps without issue.

Your mileage may vary.

**Comment by Pik-a-boo** [comment]
Just tried with Fate but with downclocking the emulated CPU to 111 mhz both instances are running at 60 FPS at first but then reduce to around 30 fps afterwards.
Maybe reducing it further could make it consistenly 30 fps.
But yeah another workaround if you dont wanna use the hack I guess (it was disabled).
@somepunkid and @ANR2ME

**Comment by somepunkid** [comment]
There may be workarounds to force a 30fps limit, but I haven't messed with much besides toggling a monitor to 30hz and using vsync, which has pretty noticeable input lag. It did work though, technically. I'm sure there's ways to do it with software, as a workaround for now (Nvidia control panel perhaps? Unsure)

**Comment by ANR2ME** [comment]
I'll remove that workaround since it's barely have any impact without force real clock sync :(

**Comment by adenovan** [comment]
downclocking the emulated cpu is the quickest work around for it as real psp also suffer from same timing issues when cross play with psvita which is the better hardware. 

this weird things on scenet adhoc library easily reproduce with bomberman games its only compatible between same cpu speed the cpu speed really affect the timing of packet sent in the timestamp and certain games has a range when its decide to drop the connection or not.

Burnout looks like normal in different playstation hardware. 

Its only my guess if the game accumulated times , some faked timestamp on get peerlist should be removed and timestamp the packet when the data arrived.

**Comment by ANR2ME** [comment]
During my investigation , the timestamps are only used as timeout calculation by comparing it with sceKernelGetSystemTimeWide.

However, some games relies on incoming data when deciding whether to render new frame not, the more data you receive than normal (ie. faster device flooding slower device buffer) the FPS/VPS.

For example, on Warriors Orochi 2 that suffers very low FPS/VPS recently, normally this game only send and receive once per frame, but during unstable FPS there could be more than 1 packets arrived and causing FPS/VPS drop for each additional packets, lowering the buffer size prevent causing the FPS/VPS to get too low with side effects of stutters due to dropped packets when buffer is full.

There are also games where the more often you send & recv the faster the FPS/VPS will be (i forgot which games).

Regarding Fate Unlimited Codes, when i compared the logs during 30 FPS and 60 FPS, there is 1 or 2 additional AdhocPollSocket while waiting for incoming data, where each AdhocPollSocket call have about 6~9ms interval.

**Comment by adenovan** [comment]
did the game log on ptp flush method ? many of game that use matching context is more broken beside of adhoc games written in old tines using pdp and ptp library only , that area need more to be explored if this game call it also , if there is a complete log i did not mind to dig the bug on real console with auto test just write what function you need the info for.

**Comment by ANR2ME** [comment]
This game does use PtpFlush during battle (which shows the 60 FPS issue), verbose logs ob SCENET only logs these 5 functions:
```
sceNetAdhocGetPtpStat(09fbf9f8, 09713900)
sceNetAdhocPollSocket(09fbfa0c, 2, 0, 1)
sceNetAdhocPtpAccept(1, [09fbfa54]=00:00:00:00:00:00, [09fbfa5a]=0, 0, 1)
sceNetAdhocPtpRecv(2,08bbb520,09fbfa5c,0,1)
sceNetAdhocPtpSend(2,08bbc060,09fbfa5c,0,1)
```

Regarding AdhocMatching, it's only used to create lobby/matchmaking, after all players ready and the game started, it's no longer used, and the game will use regular adhoc functions during game play (except GameMode which have it's own synchronization method):

PdpCreate, PtpListen, PtpAccept, PtpOpen, PtpConnect to create/open a socket and establish connection
Send/Recv to send/recv data
PtpFlush to make sure all data in send buffer are sent before waiting for a reply
PtpClose/PdpDelete to close the socket
GetPeerList/GetPeerInfo to detects timeout & disconnected players
PdpGetStat/PtpGetStat to detects available data in recv buffer, also to check for PTP connection state.
PollSocket to check whether a socket can send, recv, connect, accept, etc.

**Comment by adenovan** [comment]
roger that will do , its like a same dissidia desync issues under my investigation , will open the result after rebasing into current adhoc code

**Comment by ANR2ME** [comment]
btw @adenovan do you have a simple homebrew source which can be used for Adhocctl? 
i wanted to test something but i don't know how to initialize adhoc and adhocctl, most of the sample files from the of unofficial pspsdk i have doesn't contains adhoc example.

it would be nice if you have sample for AdhocMatching too :)

**Comment by adenovan** [comment]
@ANR2ME its on my repo psp auto test , min pspw works best for windows but you need to disable driver signature to get the full access into the shell.

No adhoc matching tdd yet beside of that adhocctl is still very manual trigger with button to test certain feature. its pretty fragile often scewlandriver crash as we using the api in unknown rule of the psp adhoc api manual , i like to trigger some driver crash so we know the exact things how its supposed to not work first and grab the full stack trace how it should not be handled that way.

Auto test has some timeout as it automated i often do write the hombrew manually just to verify the behavior of low level scenetlibrary under

**Comment by ANR2ME** [comment]
I think something really is waking up thread unexpectedly (or is this situation normal?)

Based on the Debug Logs of Space Invader Extreme (this game uses blocking PtpSend, PtpFlush, PtpRecv with infinite timeout 0)
![image](https://user-images.githubusercontent.com/7974720/96356464-1ccd4000-1119-11eb-99df-7f11bfbb5789.png)
At timestamp 54:13:101 Thread "SIE" supposed to be Blocked by PtpRecv (which switches to "user_main" thread) and Resumed with "Returning" at 54:13:103 after receiving data.

But, for some reason, not long after "SIE" was in waiting state, `sceKernelSignalSema` is waking up "SIE" thread which switch from "user_main" to "SIE" O.o Is this normal? or Did i put current thread to wait in a wrong way?

PtpRecv is blocking the thread using this code (uid = 4068)
```
int WaitBlockingAdhocSocket(...) {
...
CoreTiming::ScheduleEvent(usToCycles(100), adhocSocketNotifyEvent, threadSocketId);
__KernelWaitCurThread(WAITTYPE_NET, uid, 0, 0, false, reason);

// Fallback return value
return ERROR_NET_ADHOC_TIMEOUT;
}
```
Could someone explain why the "SIE" thread woken up even before the Scheduled Event waking it up by "Returning" a return value?

**Comment by MojoJojoDojo** [comment]
I noticed that earlier Bleach Heat The Soul games before 7 used the GameMode API as shown here : https://report.ppsspp.org/logs/kind/1093
Crazy Taxi also used it and it also suffered from being too fast... on localhost even.
Maybe a clue to why it happens on 7 and other games by Eighting (remaents of an earlier code?).
Pretty sure HTS 7 is not using GameMode as i rechecked it recently..

**Comment by ANR2ME** [comment]
Yes, HTS 7 is not using GameMode, so something else probably causing the frame to be rendered faster, may be some mutex/semaphore gets signaled too fast/early.

**Comment by MojoJojoDojo** [comment]
Interesting.
Tried out Bleach Heat The Soul 3 to find out what happens there...
It actually plays the game in correct speed on 30 fps (locally ofc).
Here is the list of games that play too fast on PPSSPP so far:
Crazy Taxi: Fare Wars (uses GameMode)
Burnout Domniator (might have changed)
Fate/Unlimited Codes
Bleach Heat The Soul 7
Need For Speed Shift (might have changed)
One of the Kamen Rider games (or all of them).

**Comment by unknownbrackets** [comment]
A lot of games rely on the vcount and vblank start funcs, and we have tests for them.  Doesn't mean they can't be wrong, but I'd be surprised if they were so obviously wrong.  It seems to me it'd have to be something more complex.

The vcount increases every vblank, so it's weird if it *waits* for it to be 2 to wait for another vblank start, you'd think that'd be waiting 3 total vblanks.  But anyway, this may indicate that it's expecting various sceNet* HLE funcs to take longer, and because they are not taking long enough, not enough vblanks have passed by the time it checks.

-[Unknown]

**Comment by unknownbrackets** [comment]
Often, really depends though.  Would have to see in the GE debugger.

-[Unknown]

**Comment by ANR2ME** [comment]
Hmm.. after the battle ends, a menu will be shown and the FPS is running at normal FPS.. may be this menu is making the FPS normal again...
![image](https://user-images.githubusercontent.com/7974720/137266202-6c34a551-2688-4e73-90b2-a7777981e2f1.png)
![image](https://user-images.githubusercontent.com/7974720/137266849-184bc942-5693-43a7-86fd-61c5eb236ce8.png)

**Comment by mainsaber** [comment]
> Hmm.. after the battle ends, a menu will be shown and the FPS is running at normal FPS.. may be this menu is making the FPS normal again... ![image](https://user-images.githubusercontent.com/7974720/137266202-6c34a551-2688-4e73-90b2-a7777981e2f1.png) ![image](https://user-images.githubusercontent.com/7974720/137266849-184bc942-5693-43a7-86fd-61c5eb236ce8.png)
> 
> Btw, why am i still seeing v1.11.3 ? while this build was based on the latest master (after 1.12.2 released)

the only problem is in the battle and this error is still not patched in the newest version of the ppsspp. in the character select screen and in the end of the battle is always in 30fps

**Comment by mainsaber** [comment]
it would be soo good if this game have rollback netcode

**Comment by mainsaber** [comment]
strangely depending on the connection of the players the fps lowers when i tested
this is dont happens in tkken 6 and dbz budokai. only with this game for some reason

**Comment by mainsaber** [comment]
> Reminder that this also happens on hardware (more on Vita) so this might not be an networking issue.. more like how the netcode was coded for the games.

strange this happening damn cavia for making this strange netcode

**Comment by mainsaber** [comment]
So theres is no way to correct the network?

**Comment by MojoJojoDojo** [comment]
> Interesting. Tried out Bleach Heat The Soul 3 to find out what happens there... It actually plays the game in correct speed on 30 fps (locally ofc). Here is the list of games that play too fast on PPSSPP so far: Crazy Taxi: Fare Wars (uses GameMode) Burnout Domniator (might have changed) Fate/Unlimited Codes Bleach Heat The Soul 7 Need For Speed Shift (might have changed) One of the Kamen Rider games (or all of them).
> 
> I am trying some games again on stable 1.11.3 to see what works so why not list them here as well.

If you use the 60 fps cheat for Crazy Taxi most elements seem to be running in proper speed.

**Comment by mainsaber** [comment]
looks like the new updates didnt solve the problem of the netcode

**Comment by Noctis-svg** [comment]
Are you still unable to play online multiplayer smoothly on Fate Unlimited Codes PSP? PSP is now at version 1.19.3.

**Comment by mainsaber** [comment]
> Are you still unable to play online multiplayer smoothly on Fate Unlimited Codes PSP? PSP is now at version 1.19.3.

i still didnt tested but looks like it still dosent work

**Comment by Noctis-svg** [comment]
> > Are you still unable to play online multiplayer smoothly on Fate Unlimited Codes PSP? PSP is now at version 1.19.3.
> 
> i still didnt tested but looks like it still dosent work

Will this game never be able to play online multiplayer smoothly?

**Comment by mainsaber** [comment]
> > > Are you still unable to play online multiplayer smoothly on Fate Unlimited Codes PSP? PSP is now at version 1.19.3.
> > 
> > 
> > i still didnt tested but looks like it still dosent work
> 
> Will this game never be able to play online multiplayer smoothly?

i dont know but i hope its better now

**Comment by mainsaber** [comment]
i was thinking, a 30fps can maybe solve this problem?

**Comment by mainsaber** [comment]
Been testing this cheat and looks like it solves the problem if anyone wants to test it out too

_C1 Online 30FPS Fix
_L 0x001F54DC 0x00000005
